torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -545,11 +625,14 @@ void Reducer::mark_variable_ready(VariableIndex index) {
       c10::OptionalStreamGuard currentStreamGuard{currentStream};
       this->finalize_backward();
       // Rebuild bucket if this is the first time to rebuild
-      if (!rebuilt_params_.empty()) {
+      if (!has_rebuilt_bucket_ && !find_unused_parameters_) {
+        TORCH_CHECK(!rebuilt_params_.empty());
+        // Unlock before rebuildBuckets() since both rebuildBuckets() and
+        // initialize_buckets() grab the lock.
+        lock.unlock();
```

Comment:
I see, yes it seems like we can have a race where multiple threads can return true for `shouldRebuildBuckets`. It's a bit tricky, since `shouldRebuildBuckets` itself should probably hold the lock since we call it from python in the join API. Maybe we can have a wrapper that simply acquires the lock.

Reply:
Changed it so that we have a single `rebuildBuckets` which checks if it should rebuild, it is a no-op if it should not rebuild. With the current design, the joined ranks will call into this no-op once if buckets don't need to be rebuilt, though.